### PR TITLE
A quadtree of pointers

### PR DESCRIPTION
**Pull request by paulmach**
This PR implements a quadtree for use with objects that implement the `geo.Pointer` interface:
```
type Pointer interface {
    Point() *Point
}
```

It supports `Find` which locates the nearest pointer to given point and `InBound` which finds all the points in the bounds. I think performance is really good:
```
BenchmarkInsert-4              	 2000000	       749 ns/op	      72 B/op	       1 allocs/op
BenchmarkInsert1000-4          	 2000000	       668 ns/op	      72 B/op	       1 allocs/op
BenchmarkFromPointer1000-4     	   10000	    197301 ns/op	  100003 B/op	       8 allocs/op
BenchmarkRandomFind1000-4      	 2000000	       907 ns/op	     112 B/op	       5 allocs/op
BenchmarkRandomInBound1000-4   	  300000	      4928 ns/op	    1828 B/op	      10 allocs/op
BenchmarkRandomInBound1000Buf-4	  500000	      3245 ns/op	      80 B/op	       4 allocs/op
```
This is based on a tree with 1000 random points. So it'll be different in practice, but I think it's pretty good. 

Still left to do:
* [x] Think about TODO that deals with how to handle close/duplicate points.
* [x] ~~Remove point~~  leave for v2
* [x] Write a README with some examples

If you have a chance. @mlerner please review.

### CODE REVIEW

File: pointers.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package geo
+
+// A Pointer is the interface for something that has a point.
+type Pointer interface {
+	// Point should return the "center" or other canonical point
+	// for the object. The caller is expected to Clone
+	// the point if changes need to be make.
+	Point() *Point
```
Comment (mlerner) [inline pointers.go]: Should this not be a pointer if we don't want to allow callers to manipulate the returned object? Maybe this question will be answered as I continue reviewing.

File: quadtree/quadtree.go [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+		return
+	}
+
+	// leaf node with a point in it.
+	nPoint := n.pointer.Point()
+
+	// TODO: parameterize this threshold, maybe 1/100th of bound width/height
+	if d := math.Abs(nPoint.X()-point.X()) + math.Abs(nPoint.Y()-point.Y()); d < 0.01 {
+		// similar/duplicate point to stop recursion.
+		n.internal = true
```
Comment (mlerner) [inline quadtree/quadtree.go]: It seems like in the rest of this function, the node is getting turned into an internal node? Could you move the setting of n.internal outside of this block?

File: quadtree/quadtree.go [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+		q.freeIndex = 0
+	}
+
+	n := &q.freeNodes[q.freeIndex]
+	q.freeIndex++
+	return n
+}
+
+func (q *Quadtree) insert(n *node, p geo.Pointer, left, right, bottom, top float64) {
+	point := p.Point()
```
Comment (mlerner) [inline quadtree/quadtree.go]: A short description of each case could provide some useful context for coming back to this code in the future.

File: quadtree/quadtree.go [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+	)
+
+	return v.closest
+}
+
+// InBound returns a slice with all the pointers in the quadtree that are
+// within the given bound. An optional buffer parameter is provided to allow
+// for the reuse of result slice memory. This function is thread safe.
+// Multiple goroutines can read from a pre-created tree.
+func (q *Quadtree) InBound(b *geo.Bound, buf ...[]geo.Pointer) []geo.Pointer {
```
Comment (mlerner) [inline quadtree/quadtree.go]: :ok_hand:

File: quadtree/quadtree.go [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+			v.Visit(n.children[1], cx, right, cy, top)
+		} else if k == 2 {
+			v.Visit(n.children[2], left, cx, bottom, cy)
+		} else if k == 3 {
+			v.Visit(n.children[3], cx, right, bottom, cy)
+		}
+	}
+}
+
+type findVisitor struct {
```
Comment (mlerner) [inline quadtree/quadtree.go]: move the visitors into their own files?

**Comment by paulmach** [comment]
@mlerner See second commit that tries the `Point() Point)`. Very clunky to use. The whole go.geo library expects pointers to points. For better or worse.

**Comment by mlerner** [comment]
It is a slightly clunky. I think it would be okay to keep the `Point` interface returning a `Point` pointer as that is the pattern that is followed elsewhere in the library.

**Comment by paulmach** [comment]
It'd be interesting to start again and see what never using a pointer for Point would look like.

Added a readme.  Please take a final look.

**Comment by mlerner** [comment]
Verbal LGTM :)

**Pull request merged**